irohad/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -46,85 +47,117 @@ namespace iroha {
-              shared_model::crypto::PublicKey(signature.publicKey()));
-          // Check chain last commit
+    void SynchronizerImpl::downloadAndApplyMissingChain(
+        const shared_model::interface::BlockVariant &committed_block_variant,
+        std::unique_ptr<ametsuchi::MutableStorage> storage) const {
+      while (true) {
+        for (const auto &signature : committed_block_variant.signatures()) {
           std::vector<std::shared_ptr<shared_model::interface::Block>> blocks;
-          network_chain.as_blocking().subscribe(
-              [&blocks](auto block) { blocks.push_back(block); });
-          auto is_chain_end_expected =
-              blocks.back()->hash() == commit_message->hash();
+          block_loader_
+              ->retrieveBlocks(
+                  shared_model::crypto::PublicKey(signature.publicKey()))
+              .as_blocking()
+              .subscribe([&blocks](auto block) { blocks.push_back(block); });
           auto chain =
               rxcpp::observable<>::iterate(blocks, rxcpp::identity_immediate());
+          // if committed block is not empty, it will be on top of downloaded
+          // chain; otherwise, it'll contain hash of top of that chain
+          auto chain_ends_with_right_block = iroha::visit_in_place(
+              committed_block_variant,
+              [last_downloaded_block = blocks.back()](
+                  std::shared_ptr<shared_model::interface::Block>
+                      committed_block) {
+                return last_downloaded_block->hash() == committed_block->hash();
+              },
+              [last_downloaded_block = blocks.back()](
+                  std::shared_ptr<shared_model::interface::EmptyBlock>
+                      committed_empty_block) {
+                return last_downloaded_block->hash()
+                    == committed_empty_block->prevHash();
+              });
 
-          if (validator_->validateChain(chain, *storage)
-              and is_chain_end_expected) {
-            // Peer send valid chain
-            mutableFactory_->commit(std::move(storage));
+          if (chain_ends_with_right_block
+              and validator_->validateChain(chain, *storage)) {
+            // peer sent valid chain
             notifier_.get_subscriber().on_next(chain);
-            // You are synchronized
+
+            for (const auto &block : blocks) {
+              // we don't need to check correctness of downloaded blocks, as
+              // it was done earlier on another peer
+              storage->apply(*block, trueStorageApplyPredicate);
+            }
+            mutable_factory_->commit(std::move(storage));
+
+            // we are finished
             return;
           }
         }
       }
     }
 
+    void SynchronizerImpl::process_commit(
+        const shared_model::interface::BlockVariant &committed_block_variant) {
+      log_->info("processing commit");
+      auto storage = createTemporaryStorage();
```

Comment:
Please consider moving it to else clause

Reply:
It's used during validation in `if`, so it must be created here